internal/services/conditionalaccess/named_location_resource.go

```diff
@@ -0,0 +1,423 @@
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Importer: tf.ValidateResourceIDPriorToImport(func(id string) error {
+			if _, err := uuid.ParseUUID(id); err != nil {
+				return fmt.Errorf("specified ID (%q) is not valid: %s", id, err)
+			}
+			return nil
+		}),
+
+		Schema: map[string]*schema.Schema{
+
+			"display_name": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateDiagFunc: validate.NoEmptyStrings,
+			},
+
+			"ip": {
+				Type:         schema.TypeList,
+				Optional:     true,
+				ForceNew:     true,
+				MaxItems:     1,
+				ExactlyOneOf: []string{"ip", "country"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"ip_ranges": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+
+						"trusted": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"country": {
+				Type:         schema.TypeList,
+				Optional:     true,
+				ForceNew:     true,
+				MaxItems:     1,
+				ExactlyOneOf: []string{"ip", "country"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"countries_and_regions": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+
+						"include_unknown_countries_and_regions": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func namedLocationResourceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ConditionalAccess.NamedLocationsClient
+
+	displayName := d.Get("display_name").(string)
+
+	base := msgraph.BaseNamedLocation{
+		DisplayName: utils.String(displayName),
+	}
+
+	if v, ok := d.GetOk("ip"); ok {
+		properties := expandIPNamedLocation(v.([]interface{}))
+		properties.BaseNamedLocation = &base
+
+		location, _, err := client.CreateIP(ctx, *properties)
+		if err != nil {
+			return tf.ErrorDiagF(err, "Could not create named location")
+		}
+		if location.ID == nil || *location.ID == "" {
+			return tf.ErrorDiagF(errors.New("Bad API response"), "Object ID returned for named location is nil/empty")
+		}
+
+		d.SetId(*location.ID)
+
+		return namedLocationResourceRead(ctx, d, meta)
+	} else if v, ok := d.GetOk("country"); ok {
+		properties := expandCountryNamedLocation(v.([]interface{}))
+		properties.BaseNamedLocation = &base
+
+		location, _, err := client.CreateCountry(ctx, *properties)
+		if err != nil {
+			return tf.ErrorDiagF(err, "Could not create named location")
+		}
+		if location.ID == nil || *location.ID == "" {
+			return tf.ErrorDiagF(errors.New("Bad API response"), "Object ID returned for named location is nil/empty")
+		}
+
+		d.SetId(*location.ID)
+
+		return namedLocationResourceRead(ctx, d, meta)
```

Comment:
This block is duplicated in both branches of the if statement. Wouldn't it make sense to move it outside ?